README.md
# mongodb-mock-server
The mock server used for testing `mongodb-core` and `node-mongodb-native`


HISTORY.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="2.0.1"></a>
## [2.0.1](https://github.com/mongodb-js/mongodb-mock-server/compare/v2.0.0...v2.0.1) (2020-03-14)


### Bug Fixes

* Revert BSON dependency fix ([4f85f7b](https://github.com/mongodb-js/mongodb-mock-server/commit/4f85f7b))
* use BSON dependency and remove BSON property ([d0aefbd](https://github.com/mongodb-js/mongodb-mock-server/commit/d0aefbd))



<a name="2.0.0"></a>
# [2.0.0](https://github.com/mongodb-js/mongodb-mock-server/compare/v1.1.0...v2.0.0) (2020-03-13)


### Bug Fixes

* remove unused bson import ([955cd4e](https://github.com/mongodb-js/mongodb-mock-server/commit/955cd4e))


### Features

* update BSON dependency to v4.0.3 ([d756b10](https://github.com/mongodb-js/mongodb-mock-server/commit/d756b10))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/mongodb-js/mongodb-mock-server/compare/v1.0.1...v1.1.0) (2020-01-09)


### Features

* support running mock server with TLS connections ([4c86be3](https://github.com/mongodb-js/mongodb-mock-server/commit/4c86be3))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/mongodb-js/mongodb-mock-server/compare/v1.0.0...v1.0.1) (2019-02-06)


### Features

* **ismaster:** add mock server identifier to default ismaster ([c77dddf](https://github.com/mongodb-js/mongodb-mock-server/commit/c77dddf))



<a name="1.0.0"></a>
# 1.0.0 (2018-01-06)


### Features

* **mock-server:** initial code dump of mock server implementation ([2d340a9](https://github.com/mongodb-js/mongodb-mock-server/commit/2d340a9))


test/basic.js
'use strict';
const mock = require('..');

describe('MockServer', function() {
  afterEach(() => mock.cleanup());

  it('should work', function() {
    mock.createServer().then(server => {
      server.setMessageHandler(req => req.reply({ ok: 1 }));
    });
  });
});


index.js
'use strict';
const fs = require('fs');
const MockServer = require('./lib/server');

let mockServers = [];
const cleanup = (spy, callback) => {
  if (typeof spy === 'function') {
    callback = spy;
    spy = undefined;
  }

  if (spy) {
    const alreadyDrained = spy.connectionCount() === 0;
    const drainedPromise = !alreadyDrained
      ? new Promise(resolve => spy.once('drained', () => resolve()))
      : Promise.resolve();

    const cleanupPromise = Promise.all(mockServers.map(server => server.destroy()))
      .then(drainedPromise)
      .then(() => {
        mockServers = [];
      })
      .catch(err => {
        mockServers = [];
        throw err;
      });

    if (typeof callback !== 'function') {
      return cleanupPromise;
    }

    return cleanupPromise.then(() => callback(null, null)).catch(err => callback(err, null));
  } else {
    const cleanupPromise = Promise.all(mockServers.map(server => server.destroy()));
    if (typeof callback !== 'function') {
      return cleanupPromise;
    }

    return cleanupPromise.then(() => callback(null, null)).catch(err => callback(err, null));
  }
};

// Default message fields
const DEFAULT_ISMASTER = {
  __nodejs_mock_server__: true,
  ismaster: true,
  maxBsonObjectSize: 16777216,
  maxMessageSizeBytes: 48000000,
  maxWriteBatchSize: 1000,
  localTime: new Date(),
  maxWireVersion: 5,
  minWireVersion: 0,
  ok: 1
};

const DEFAULT_ISMASTER_36 = Object.assign({}, DEFAULT_ISMASTER, {
  maxWireVersion: 6,
  logicalSessionTimeoutMinutes: 10
});

/*
 * Main module
 */
module.exports = {
  createServer: function(port, host, options) {
    port = port || 0;
    host = host || 'localhost';
    options = options || {};

    if (process.env.MONGODB_SERVER_PEM || process.env.MONGODB_CA_PEM) {
      if (process.env.MONGODB_SERVER_PEM == null) {
        throw new Error('MONGODB_SERVER_PEM must be provided for TLS support');
      }

      if (process.env.MONGODB_CA_PEM == null) {
        throw new Error('MONGODB_CA_PEM must be provided for TLS support');
      }

      Object.assign(options, {
        tls: true,
        ca: fs.readFileSync(process.env.MONGODB_CA_PEM),
        cert: fs.readFileSync(process.env.MONGODB_SERVER_PEM),
        key: fs.readFileSync(process.env.MONGODB_SERVER_PEM)
      });
    }

    let mockServer = new MockServer(port, host, options);
    mockServers.push(mockServer);
    return mockServer.start();
  },

  cleanup: cleanup,
  DEFAULT_ISMASTER: DEFAULT_ISMASTER,
  DEFAULT_ISMASTER_36: DEFAULT_ISMASTER_36
};


lib/request.js
'use strict';
const Long = require('bson').Long;
const snappy = require('snappy');
const zlib = require('zlib');
const opcodes = require('./utils').opcodes;
const compressorIDs = require('./utils').compressorIDs;

/*
 * Request class
 */
const Request = function(server, connection, response) {
  this.server = server;
  this.connection = connection;
  this.response = response;
  this.bson = server.bson;
};

Request.prototype.receive = function() {
  return Promise.resolve();
};

Request.prototype.reply = function(documents, options) {
  options = options || {};
  documents = Array.isArray(documents) ? documents : [documents];

  // Unpack any variables we need
  let cursorId = options.cursorId || Long.ZERO;
  let responseFlags = typeof options.responseFlags === 'number' ? options.responseFlags : 0;
  let startingFrom = typeof options.startingFrom === 'number' ? options.startingFrom : 0;
  let numberReturned = documents.length;

  // Additional response Options
  let killConnectionAfterNBytes =
    typeof options.killConnectionAfterNBytes === 'number'
      ? options.killConnectionAfterNBytes
      : null;

  // Create the Response document
  let response;
  if (options.compression) {
    response = new CompressedResponse(
      this.bson,
      {
        cursorId: cursorId,
        responseFlags: responseFlags,
        startingFrom: startingFrom,
        numberReturned: numberReturned,
        documents: documents
      },
      {
        // Header field
        responseTo: this.response.requestId,
        requestId: this.response.requestId + 1,
        originalOpCode: options.originalOpCode,
        compressorID: compressorIDs[options.compression.compressor] || 0
      }
    );
  } else {
    response = new Response(this.bson, documents, {
      // Header field
      responseTo: this.response.requestId,
      requestId: this.response.requestId + 1,

      // The OP_REPLY message field
      cursorId: cursorId,
      responseFlags: responseFlags,
      startingFrom: startingFrom,
      numberReturned: numberReturned
    });
  }

  // Get the buffers
  let buffer = response.toBin();

  // Do we kill connection after n bytes
  if (killConnectionAfterNBytes == null) {
    this.connection.write(buffer);
  } else {
    // Fail to send whole reply
    if (killConnectionAfterNBytes <= buffer.length) {
      this.connection.write(buffer.slice(0, killConnectionAfterNBytes));
      this.connection.destroy();
    }
  }
};

Object.defineProperty(Request.prototype, 'type', {
  get: function() {
    return this.response.type;
  }
});

Object.defineProperty(Request.prototype, 'document', {
  get: function() {
    return this.response.documents[0];
  }
});

const Response = function(bson, documents, options) {
  this.bson = bson;
  // Header
  this.requestId = options.requestId;
  this.responseTo = options.responseTo;
  this.opCode = 1;

  // Message fields
  (this.cursorId = options.cursorId),
    (this.responseFlags = options.responseFlags),
    (this.startingFrom = options.startingFrom),
    (this.numberReturned = options.numberReturned);

  // Store documents
  this.documents = documents;
};

/**
 * @ignore
 * Preparing a compressed response of the OP_COMPRESSED type
 */
const CompressedResponse = function(bson, uncompressedResponse, options) {
  this.bson = bson;

  // Header
  this.requestId = options.requestId;
  this.responseTo = options.responseTo;
  this.opCode = opcodes.OP_COMPRESSED;

  // OP_COMPRESSED fields
  this.originalOpCode = opcodes.OP_REPLY;
  this.compressorID = options.compressorID;

  this.uncompressedResponse = {
    cursorId: uncompressedResponse.cursorId,
    responseFlags: uncompressedResponse.responseFlags,
    startingFrom: uncompressedResponse.startingFrom,
    numberReturned: uncompressedResponse.numberReturned,
    documents: uncompressedResponse.documents
  };
};

Response.prototype.toBin = function() {
  let self = this;
  let buffers = [];

  // Serialize all the docs
  let docs = this.documents.map(function(x) {
    return self.bson.serialize(x);
  });

  // Document total size
  let docsSize = 0;
  docs.forEach(function(x) {
    docsSize = docsSize + x.length;
  });

  // Calculate total size
  let totalSize =
    4 +
    4 +
    4 +
    4 + // Header size
    4 +
    8 +
    4 +
    4 + // OP_REPLY Header size
    docsSize; // OP_REPLY Documents

  // Header and op_reply fields
  let header = new Buffer(4 + 4 + 4 + 4 + 4 + 8 + 4 + 4);

  // Write total size
  writeInt32(header, 0, totalSize);
  // Write requestId
  writeInt32(header, 4, this.requestId);
  // Write responseId
  writeInt32(header, 8, this.responseTo);
  // Write opcode
  writeInt32(header, 12, this.opCode);
  // Write responseflags
  writeInt32(header, 16, this.responseFlags);
  // Write cursorId
  writeInt64(header, 20, this.cursorId);
  // Write startingFrom
  writeInt32(header, 28, this.startingFrom);
  // Write startingFrom
  writeInt32(header, 32, this.numberReturned);

  // Add header to the list of buffers
  buffers.push(header);
  // Add docs to list of buffers
  buffers = buffers.concat(docs);
  // Return all the buffers
  return Buffer.concat(buffers);
};

CompressedResponse.prototype.toBin = function() {
  let self = this;
  let buffers = [];

  // Serialize all the docs
  let docs = this.uncompressedResponse.documents.map(function(x) {
    return self.bson.serialize(x);
  });

  // Document total size
  let uncompressedSize = 4 + 8 + 4 + 4; // OP_REPLY Header size
  docs.forEach(function(x) {
    uncompressedSize = uncompressedSize + x.length;
  });

  let dataToBeCompressedHeader = new Buffer(20);
  let dataToBeCompressedBody = Buffer.concat(docs);

  // Write response flags
  writeInt32(dataToBeCompressedHeader, 0, this.uncompressedResponse.responseFlags);
  writeInt64(dataToBeCompressedHeader, 4, this.uncompressedResponse.cursorId);
  writeInt32(dataToBeCompressedHeader, 12, this.uncompressedResponse.startingFrom);
  writeInt32(dataToBeCompressedHeader, 16, this.uncompressedResponse.numberReturned);

  let dataToBeCompressed = Buffer.concat([dataToBeCompressedHeader, dataToBeCompressedBody]);

  // Compress the data
  let compressedData;
  switch (this.compressorID) {
    case compressorIDs.snappy:
      compressedData = snappy.compressSync(dataToBeCompressed);
      break;
    case compressorIDs.zlib:
      compressedData = zlib.deflateSync(dataToBeCompressed);
      break;
    default:
      compressedData = dataToBeCompressed;
  }

  // Calculate total size
  let totalSize =
    4 +
    4 +
    4 +
    4 + // Header size
    4 +
    4 +
    1 + // OP_COMPRESSED fields
    compressedData.length; // OP_REPLY fields

  // Header and op_reply fields
  let header = new Buffer(totalSize - compressedData.length);

  // Write total size
  writeInt32(header, 0, totalSize);
  // Write requestId
  writeInt32(header, 4, this.requestId);
  // Write responseId
  writeInt32(header, 8, this.responseTo);
  // Write opcode
  writeInt32(header, 12, this.opCode);
  // Write original opcode`
  writeInt32(header, 16, this.originalOpCode);
  // Write uncompressed message size
  writeInt64(header, 20, Long.fromNumber(uncompressedSize));
  // Write compressorID
  header[24] = this.compressorID & 0xff;

  // Add header to the list of buffers
  buffers.push(header);
  // Add docs to list of buffers
  buffers = buffers.concat(compressedData);

  return Buffer.concat(buffers);
};

const writeInt32 = function(buffer, index, value) {
  buffer[index] = value & 0xff;
  buffer[index + 1] = (value >> 8) & 0xff;
  buffer[index + 2] = (value >> 16) & 0xff;
  buffer[index + 3] = (value >> 24) & 0xff;
  return;
};

const writeInt64 = function(buffer, index, value) {
  let lowBits = value.getLowBits();
  let highBits = value.getHighBits();
  // Encode low bits
  buffer[index] = lowBits & 0xff;
  buffer[index + 1] = (lowBits >> 8) & 0xff;
  buffer[index + 2] = (lowBits >> 16) & 0xff;
  buffer[index + 3] = (lowBits >> 24) & 0xff;
  // Encode high bits
  buffer[index + 4] = highBits & 0xff;
  buffer[index + 5] = (highBits >> 8) & 0xff;
  buffer[index + 6] = (highBits >> 16) & 0xff;
  buffer[index + 7] = (highBits >> 24) & 0xff;
  return;
};

module.exports = Request;


lib/protocol.js
'use strict';
const BSON = require('bson');

// Reads in a C style string
const readCStyleStringSpecial = function(buffer, index) {
  // Get the start search index
  let i = index;
  // Locate the end of the c string
  while (buffer[i] !== 0x00 && i < buffer.length) {
    i++;
  }
  // If are at the end of the buffer there is a problem with the document
  if (i >= buffer.length) throw new Error('Bad BSON Document: illegal CString');
  // Grab utf8 encoded string
  const string = buffer.toString('utf8', index, i);
  // Update index position
  index = i + 1;
  // Return string
  return { s: string, i: index };
};

const Query = function(data) {
  // The type of message
  this.type = 'op_query';
  // The number of documents
  this.documents = [];
  // Unpack the message
  let index = 0;
  // Message size
  this.size =
    data[index] | (data[index + 1] << 8) | (data[index + 2] << 16) | (data[index + 3] << 24);
  index = index + 4;
  // requestId
  this.requestId =
    data[index] | (data[index + 1] << 8) | (data[index + 2] << 16) | (data[index + 3] << 24);
  index = index + 4;
  // responseTo
  this.responseTo =
    data[index] | (data[index + 1] << 8) | (data[index + 2] << 16) | (data[index + 3] << 24);
  index = index + 4;
  // opCode
  this.opCode =
    data[index] | (data[index + 1] << 8) | (data[index + 2] << 16) | (data[index + 3] << 24);
  index = index + 4;

  // flags
  this.flags =
    data[index] | (data[index + 1] << 8) | (data[index + 2] << 16) | (data[index + 3] << 24);
  index = index + 4;

  // Read the full collection name
  const result = readCStyleStringSpecial(data, index);
  this.ns = result.s;
  index = result.i;

  // numberToSkip
  this.numberToSkip =
    data[index] | (data[index + 1] << 8) | (data[index + 2] << 16) | (data[index + 3] << 24);
  index = index + 4;

  // numberToReturn
  this.numberToReturn =
    data[index] | (data[index + 1] << 8) | (data[index + 2] << 16) | (data[index + 3] << 24);
  index = index + 4;

  // Read the document size
  let docSize =
    data[index] | (data[index + 1] << 8) | (data[index + 2] << 16) | (data[index + 3] << 24);

  // Deserialize the document
  this.documents.push(BSON.deserialize(data.slice(index, index + docSize)));
  index = index + docSize;

  // No field selection
  if (index === data.length) {
    return;
  }

  // Read the projection document size
  docSize =
    data[index] | (data[index + 1] << 8) | (data[index + 2] << 16) | (data[index + 3] << 24);
  this.projection = BSON.deserialize(data.slice(index, index + docSize));
};

const GetMore = function() {};
const KillCursor = function() {};
const Insert = function() {};
const Update = function() {};
const Delete = function() {};

module.exports = {
  Query: Query,
  GetMore: GetMore,
  KillCursor: KillCursor,
  Insert: Insert,
  Update: Update,
  Delete: Delete
};


lib/server.js
'use strict';
const net = require('net');
const tls = require('tls');
const Snappy = require('snappy');
const zlib = require('zlib');
const MESSAGE_HEADER_SIZE = require('./utils').MESSAGE_HEADER_SIZE;
const opcodes = require('./utils').opcodes;
const compressorIDs = require('./utils').compressorIDs;
const Request = require('./request');
const Query = require('./protocol').Query;
const GetMore = require('./protocol').GetMore;
const KillCursor = require('./protocol').KillCursor;
const Insert = require('./protocol').Insert;
const Update = require('./protocol').Update;
const Delete = require('./protocol').Delete;
const EventEmitter = require('events');

/*
 * MockServer class
 */
class MockServer extends EventEmitter {
  constructor(port, host, options) {
    super();

    // Special handlers
    options = options || {};

    // Do we have an onRead function
    this.onRead = typeof options.onRead === 'function' ? options.onRead : null;

    // Save the settings
    this.host = host;
    this.port = port;

    // Create a server socket
    this.server = options.tls ? tls.createServer(options) : net.createServer(options);
    this.tlsEnabled = !!options.tls;

    // Responses
    this.messages = [];

    // state
    this.state = 'stopped';

    // Number of connections
    this.connections = 0;

    // sockets
    this.sockets = [];

    // message handlers
    this.messageHandlers = Object.create(null);
  }

  /**
   *
   */
  uri() {
    return `${this.host}:${this.port}`;
  }

  /**
   *
   */
  address() {
    return { host: this.host, port: this.port };
  }

  /**
   *
   */
  destroy() {
    const self = this;
    if (self.state === 'destroyed') {
      return Promise.resolve();
    }

    return new Promise((resolve, reject) => {
      self.sockets.forEach(socket => socket.destroy());
      self.server.close(err => {
        if (err) return reject(err);
        self.state = 'destroyed';
        resolve();
      });
    });
  }

  /**
   *
   */
  start() {
    const self = this;
    return new Promise((resolve, reject) => {
      self.server.on('error', err => {
        console.log('!!!!!!!!!!!!!!!!!!!! error reject');
        reject(err);
      });

      const connectionEventName = self.tlsEnabled ? 'secureConnection' : 'connection';
      self.server.on(connectionEventName, c => {
        self.connections = self.connections + 1;
        self.sockets.push(c);

        c.on('error', e => {
          console.warn('connection error: ', e);
        });

        c.on(
          'data',
          dataHandler(
            self,
            {
              buffer: new Buffer(0),
              stubBuffer: new Buffer(0),
              sizeOfMessage: 0,
              bytesRead: 0,
              maxBsonMessageSize: 1024 * 1024 * 48
            },
            c
          )
        );

        c.on('close', () => {
          self.connections = self.connections - 1;
          const index = self.sockets.indexOf(c);

          if (index !== -1) {
            self.sockets.splice(index, 1);
          }
        });
      });

      self.server.listen(self.port, self.host, () => {
        // update address information if necessary
        self.host = self.server.address().address;
        self.port = self.server.address().port;

        resolve(self);
      });

      self.on('message', function(message, connection) {
        const request = new Request(self, connection, message);
        if (self.genericMessageHandler) {
          try {
            self.genericMessageHandler(request);
          } catch (err) {
            console.log(err.stack);
          }

          return;
        }

        const command = Object.keys(request.document)[0];
        if (self.messageHandlers[command]) {
          let messageHandler = self.messageHandlers[command];
          if (Array.isArray(messageHandler)) {
            if (messageHandler.length === 0) {
              delete self.handlers[command];
            } else {
              messageHandler = messageHandler.shift();
            }
          }

          try {
            messageHandler(request);
          } catch (err) {
            console.log(err.stack);
          }
        } else {
          self.messages.push(request);
        }
      });

      self.state = 'running';
    });
  }

  /**
   *
   */
  receive() {
    const self = this;
    return new Promise((resolve, reject) => {
      const waiting = () => {
        if (self.state === 'destroyed') {
          return reject(new Error('mock server is in destroyed state'));
        }

        // If we have a message return it
        if (self.messages.length > 0) {
          const message = self.messages.shift();
          return resolve(message);
        }

        setTimeout(waiting, 10);
      };

      waiting();
    });
  }

  /**
   * Legacy method for registering a message handler. This method allows for setting a
   * generic message handler, if no command type is specified.
   *
   * @param {string|function} typeOrHandler the type of command to register a handler for, or the generic handler
   * @param {function} [messageHandler] the optional message handler, if a type was specified
   */
  setMessageHandler(type, messageHandler) {
    if (typeof type === 'function') (messageHandler = type), (type = undefined);

    if (type == null) {
      this.genericMessageHandler = messageHandler;
      return;
    }

    this.messageHandlers[type] = messageHandler;
  }

  /**
   * Adds a message handler to the mock server, optionally adding it to an array of
   * handlers.
   *
   * @param {string} type the command type to register this handler for
   * @param {function} messageHandler the handler for the message
   * @returns {MockServer} the mock server
   */
  addMessageHandler(type, messageHandler) {
    if (this.messageHandlers[type]) {
      if (Array.isArray(this.messageHandlers[type])) {
        this.messageHandlers[type].push(messageHandler);
      } else {
        this.messageHandlers[type] = [messageHandler, this.messageHandlers[type]];
      }
    } else {
      this.messageHandlers[type] = messageHandler;
    }

    return this;
  }
}

const protocol = function(self, message) {
  let index = 0;
  self.isCompressed = false;
  // Get the size for the message
  const size =
    message[index++] |
    (message[index++] << 8) |
    (message[index++] << 16) |
    (message[index++] << 24);
  if (size !== message.length) throw new Error('corrupt wire protocol message');
  // Adjust to opcode
  index = 12;
  // Get the opCode for the message
  let type =
    message[index++] |
    (message[index++] << 8) |
    (message[index++] << 16) |
    (message[index++] << 24);

  // Unpack and decompress if the message is OP_COMPRESSED
  if (type === opcodes.OP_COMPRESSED) {
    const requestID = message.readInt32LE(4);
    const responseTo = message.readInt32LE(8);
    const originalOpcode = message.readInt32LE(16);
    const uncompressedSize = message.readInt32LE(20);
    const compressorID = message.readUInt8(24);

    const compressedData = message.slice(25);
    let uncompressedData;
    switch (compressorID) {
      case compressorIDs.snappy:
        uncompressedData = Snappy.uncompressSync(compressedData);
        break;
      case compressorIDs.zlib:
        uncompressedData = zlib.inflateSync(compressedData);
        break;
      default:
        uncompressedData = compressedData;
    }

    if (uncompressedData.length !== uncompressedSize) {
      throw new Error(
        'corrupt wire protocol message: uncompressed message is not the correct size'
      );
    }

    // Reconstruct the msgHeader of the uncompressed opcode
    const newMsgHeader = Buffer(MESSAGE_HEADER_SIZE);
    newMsgHeader.writeInt32LE(MESSAGE_HEADER_SIZE + uncompressedData.length, 0);
    newMsgHeader.writeInt32LE(requestID, 4);
    newMsgHeader.writeInt32LE(responseTo, 8);
    newMsgHeader.writeInt32LE(originalOpcode, 12);

    // Full uncompressed message
    message = Buffer.concat([newMsgHeader, uncompressedData]);
    type = originalOpcode;

    // Compressed flag
    self.isCompressed = true;
  }

  // Switch on type
  if (type === opcodes.OP_UPDATE) return new Update(message);
  if (type === opcodes.OP_INSERT) return new Insert(message);
  if (type === opcodes.OP_QUERY) return new Query(message);
  if (type === opcodes.OP_GETMORE) return new GetMore(message);
  if (type === opcodes.OP_DELETE) return new Delete(message);
  if (type === opcodes.OP_KILL_CURSORS) return new KillCursor(message);
  throw new Error('unknown wire protocol message type');
};

const dataHandler = function(server, self, connection) {
  return function(data) {
    // Parse until we are done with the data
    while (data.length > 0) {
      // Call the onRead function
      if (typeof server.onRead === 'function') {
        // If onRead returns true, terminate the reading for this connection as
        // it's dead
        if (server.onRead(server, connection, self.buffer, self.bytesRead)) {
          break;
        }
      }

      // If we still have bytes to read on the current message
      if (self.bytesRead > 0 && self.sizeOfMessage > 0) {
        // Calculate the amount of remaining bytes
        let remainingBytesToRead = self.sizeOfMessage - self.bytesRead;
        // Check if the current chunk contains the rest of the message
        if (remainingBytesToRead > data.length) {
          // Copy the new data into the exiting buffer (should have been allocated when we know the message size)
          data.copy(self.buffer, self.bytesRead);
          // Adjust the number of bytes read so it point to the correct index in the buffer
          self.bytesRead = self.bytesRead + data.length;

          // Reset state of buffer
          data = new Buffer(0);
        } else {
          // Copy the missing part of the data into our current buffer
          data.copy(self.buffer, self.bytesRead, 0, remainingBytesToRead);
          // Slice the overflow into a new buffer that we will then re-parse
          data = data.slice(remainingBytesToRead);

          // Emit current complete message
          try {
            let emitBuffer = self.buffer;
            // Reset state of buffer
            self.buffer = null;
            self.sizeOfMessage = 0;
            self.bytesRead = 0;
            self.stubBuffer = null;
            // Emit the buffer
            server.emit('message', protocol(server, emitBuffer), connection);
          } catch (err) {
            let errorObject = {
              err: 'socketHandler',
              trace: err,
              bin: self.buffer,
              parseState: {
                sizeOfMessage: self.sizeOfMessage,
                bytesRead: self.bytesRead,
                stubBuffer: self.stubBuffer
              }
            };
            // We got a parse Error fire it off then keep going
            server.emit('parseError', errorObject, self);
          }
        }
      } else {
        // Stub buffer is kept in case we don't get enough bytes to determine the
        // size of the message (< 4 bytes)
        if (self.stubBuffer != null && self.stubBuffer.length > 0) {
          // If we have enough bytes to determine the message size let's do it
          if (self.stubBuffer.length + data.length > 4) {
            // Prepad the data
            let newData = new Buffer(self.stubBuffer.length + data.length);
            self.stubBuffer.copy(newData, 0);
            data.copy(newData, self.stubBuffer.length);
            // Reassign for parsing
            data = newData;

            // Reset state of buffer
            self.buffer = null;
            self.sizeOfMessage = 0;
            self.bytesRead = 0;
            self.stubBuffer = null;
          } else {
            // Add the the bytes to the stub buffer
            let newStubBuffer = new Buffer(self.stubBuffer.length + data.length);
            // Copy existing stub buffer
            self.stubBuffer.copy(newStubBuffer, 0);
            // Copy missing part of the data
            data.copy(newStubBuffer, self.stubBuffer.length);
            // Exit parsing loop
            data = new Buffer(0);
          }
        } else {
          if (data.length > 4) {
            // Retrieve the message size
            let sizeOfMessage = data[0] | (data[1] << 8) | (data[2] << 16) | (data[3] << 24);
            // If we have a negative sizeOfMessage emit error and return
            if (sizeOfMessage < 0 || sizeOfMessage > self.maxBsonMessageSize) {
              let errorObject = {
                err: 'socketHandler',
                trace: '',
                bin: self.buffer,
                parseState: {
                  sizeOfMessage: sizeOfMessage,
                  bytesRead: self.bytesRead,
                  stubBuffer: self.stubBuffer
                }
              };
              // We got a parse Error fire it off then keep going
              server.emit('parseError', errorObject, self);
              return;
            }

            // Ensure that the size of message is larger than 0 and less than the max allowed
            if (
              sizeOfMessage > 4 &&
              sizeOfMessage < self.maxBsonMessageSize &&
              sizeOfMessage > data.length
            ) {
              self.buffer = new Buffer(sizeOfMessage);
              // Copy all the data into the buffer
              data.copy(self.buffer, 0);
              // Update bytes read
              self.bytesRead = data.length;
              // Update sizeOfMessage
              self.sizeOfMessage = sizeOfMessage;
              // Ensure stub buffer is null
              self.stubBuffer = null;
              // Exit parsing loop
              data = new Buffer(0);
            } else if (
              sizeOfMessage > 4 &&
              sizeOfMessage < self.maxBsonMessageSize &&
              sizeOfMessage === data.length
            ) {
              try {
                let emitBuffer = data;
                // Reset state of buffer
                self.buffer = null;
                self.sizeOfMessage = 0;
                self.bytesRead = 0;
                self.stubBuffer = null;
                // Exit parsing loop
                data = new Buffer(0);
                // Emit the message
                server.emit('message', protocol(server, emitBuffer), connection);
              } catch (err) {
                let errorObject = {
                  err: 'socketHandler',
                  trace: err,
                  bin: self.buffer,
                  parseState: {
                    sizeOfMessage: self.sizeOfMessage,
                    bytesRead: self.bytesRead,
                    stubBuffer: self.stubBuffer
                  }
                };
                // We got a parse Error fire it off then keep going
                server.emit('parseError', errorObject, self);
              }
            } else if (sizeOfMessage <= 4 || sizeOfMessage > self.maxBsonMessageSize) {
              let errorObject = {
                err: 'socketHandler',
                trace: null,
                bin: data,
                parseState: {
                  sizeOfMessage: sizeOfMessage,
                  bytesRead: 0,
                  buffer: null,
                  stubBuffer: null
                }
              };
              // We got a parse Error fire it off then keep going
              server.emit('parseError', errorObject, self);

              // Clear out the state of the parser
              self.buffer = null;
              self.sizeOfMessage = 0;
              self.bytesRead = 0;
              self.stubBuffer = null;
              // Exit parsing loop
              data = new Buffer(0);
            } else {
              let emitBuffer = data.slice(0, sizeOfMessage);
              // Reset state of buffer
              self.buffer = null;
              self.sizeOfMessage = 0;
              self.bytesRead = 0;
              self.stubBuffer = null;
              // Copy rest of message
              data = data.slice(sizeOfMessage);
              // Emit the message
              server.emit('message', protocol(server, emitBuffer), connection);
            }
          } else {
            // Create a buffer that contains the space for the non-complete message
            self.stubBuffer = new Buffer(data.length);
            // Copy the data to the stub buffer
            data.copy(self.stubBuffer, 0);
            // Exit parsing loop
            data = new Buffer(0);
          }
        }
      }
    }
  };
};

module.exports = MockServer;
